src/main/java/software/amazon/smithy/lsp/project/ProjectDependencyResolver.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.lsp.project;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import software.amazon.smithy.build.SmithyBuild;
+import software.amazon.smithy.build.model.MavenConfig;
+import software.amazon.smithy.build.model.MavenRepository;
+import software.amazon.smithy.cli.EnvironmentVariable;
+import software.amazon.smithy.cli.dependencies.DependencyResolver;
+import software.amazon.smithy.cli.dependencies.MavenDependencyResolver;
+import software.amazon.smithy.cli.dependencies.ResolvedArtifact;
+import software.amazon.smithy.lsp.util.Result;
+
+/**
+ * Resolves all Maven dependencies and {@link ProjectDependency} for a project.
+ *
+ * <p>Resolving a {@link ProjectDependency} is as simple as getting its path
+ * relative to the project root, but is done here in order to be loaded the
+ * same way as Maven dependencies.
+ */
+final class ProjectDependencyResolver {
+    // Taken from smithy-cli ConfigurationUtils
+    private static final Supplier<MavenRepository> CENTRAL = () -> MavenRepository.builder()
+            .url("https://repo.maven.apache.org/maven2")
+            .build();
+
+    private ProjectDependencyResolver() {
+    }
+
+    static Result<List<Path>, Exception> resolveDependencies(Path root, ProjectConfig config) {
+        return Result.ofFallible(() -> {
+            List<Path> deps = ProjectDependencyResolver.create(config).resolve()
+                    .stream()
+                    .map(ResolvedArtifact::getPath)
+                    .collect(Collectors.toCollection(ArrayList::new));
+            config.getDependencies().forEach((projectDependency) -> {
+                // TODO: Not sure if this needs to check for existence
+                Path path = root.resolve(projectDependency.getPath()).normalize();
+                deps.add(path);
+            });
+            return deps;
+        });
+    }
+
+    // Taken (roughly) from smithy-cli ClasspathAction::resolveDependencies
+    private static DependencyResolver create(ProjectConfig config) {
+        // DependencyResolver delegate = new MavenDependencyResolver(EnvironmentVariable.SMITHY_MAVEN_CACHE.get());
```

Comment:
Mentioned it in the comment [here](https://github.com/smithy-lang/smithy-language-server/pull/146#discussion_r1657908532), but I've been having the server go based on source files, and ignoring any artifacts created by the CLI or other build systems. In this case, I did it so the server doesn't write to the cache, and yes, so whenever smithy-build.json changes, the project resolver refreshes. The cache benefits the CLI more because it needs it on every command, whereas this only cares when smithy-build.json changes. That could be further improved to check for when deps actually changed in smithy-build.json (there's comments elsewhere about granular project updates based on smithy-build.json changes)